new-contrib: Audio Whisper API with Local Device Microphones

CarlKho-Minerva opened the pull request:
### Summary

This PR adds a new notebook that demonstrates how to use the Whisper API to transcribe text from your device's microphone. The notebook includes steps to record audio, transcribe it using the Whisper API, and copy the transcription to the clipboard. It aims to provide a practical guide for users who want to integrate speech-to-text functionality into their applications.

*This pull request was written by Chat GPT and reviewed by a human. **The article, however, is made by a human.**

### Motivation

This tutorial was created because the functionality to transcribe speech to text from a microphone is not well-documented. I found the mic speech-to-text option in the ChatGPT apps (not websites) extremely helpful for day-to-day operations and wanted to save others from having to learn about different audio processing modules.

### For new content

When contributing new content, read through our contribution guidelines, and mark the following action items as completed:

- [x] I have added a new entry in registry.yaml (and, optionally, in authors.yaml) so that my content renders on the cookbook website.
- [x] I have conducted a self-review of my content based on the contribution guidelines (my previous PR message detailed on every one of these 😅):
  - [x] **Relevance:** This content is related to building with OpenAI technologies and is useful to others.
  - [x] **Uniqueness:** I have searched for related examples in the OpenAI Cookbook and verified that my content offers new insights or unique information compared to existing documentation.
  - [x] **Spelling and Grammar:** I have checked for spelling or grammatical mistakes.
  - [x] **Clarity:** I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [x] **Correctness:** The information I include is correct, and all of my code executes successfully.
  - [x] **Completeness:** I have explained everything fully, including all necessary references and citations.

CarlKho-Minerva committed:
new-contrib: submission - Audio Whisper API with Device Microphones

CarlKho-Minerva committed:
chore: Updated yaml info + and typo correcting

CarlKho-Minerva commented:
My previous PR message before I updated. It's mostly justification for each criteria.

#### Introduction:
This contribution introduces a practical guide on using the Whisper API to transcribe audio from a device's microphone. The notebook includes steps to record audio, transcribe it using the Whisper API, and copy the transcription to the clipboard, providing an accessible and useful resource for AI builders.

#### Justification:

**1. Relevance:**
This guide is relevant as it utilizes OpenAI's Whisper API, allowing users to transcribe audio directly from their devices. This functionality aligns with OpenAI's mission to provide practical applications of AI technologies.

**2. Usefulness:**
The contribution is highly useful for developers who need a reliable method to convert speech to text. It can be used in various applications, such as real-time transcription services, voice command systems, and accessibility tools for individuals with hearing impairments. I found the mic speech-to-text option in the ChatGPT apps (not websites) very helpful for day-to-day operations and wanted to extend this functionality.

**3. Uniqueness:**
While there are existing examples of using the Whisper API, this notebook uniquely combines multiple functionalities—recording audio, transcribing it, and copying the transcription to the clipboard—into one cohesive guide. This integration simplifies the process for users and provides a complete solution in a single resource. Given that this functionality isn't extensively documented yet, I believe this tutorial can fill an important gap.

**4. Clarity:**
The notebook is written in clear, easy-to-understand language, with step-by-step instructions and code snippets. It includes detailed comments and explanations, making it accessible even to beginners.

**5. Correctness:**
The code has been tested and verified for accuracy. It includes all necessary imports and setup instructions, ensuring that users can replicate the process without errors.

**6. Conciseness:**
The guide is concise yet thorough, covering all essential steps without unnecessary information. It is structured to provide maximum value in a compact format.

**7. Completeness:**
The contribution is complete, covering everything from setting up microphone permissions to troubleshooting common issues. It provides all necessary context and resources, ensuring users have a comprehensive understanding of the process.

**8. Grammar:**
The notebook is free from grammatical and spelling errors, ensuring professional quality and readability.

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva commented:
Hello @justonf @ted-at-openai and other people relevant to the review process.

I would like to gently request for feedback and ask if there has been something that my contribution has failed to do. I will immediately correct them if ever.

Looking forward to your response. Thank you!

CarlKho-Minerva commented:
Please feel free to give me quick feedback so I may improve it soon. If any. :)

| Criteria | Description | Score |
|----------|-------------|-------|
| Relevance | Is the content related to building with OpenAI technologies? Is it useful to others? |  |
| Uniqueness | Does the content offer new insights or unique information compared to existing documentation? |  |
| Clarity | Is the language easy to understand? Are things well-explained? Is the title clear? |  |
| Correctness | Are the facts, code snippets, and examples correct and reliable? Does everything execute correctly? |  |
| Conciseness | Is the content concise? Are all details necessary? Can it be made shorter? |  |
| Completeness | Is the content thorough and detailed? Are there things that weren’t explained fully? |  |
| Grammar | Are there grammatical or spelling errors present |  |

CarlKho-Minerva commented:
gently bumping this up. willing to revise and have learned a lot about dealing after submitting [my hackathon entry](https://youtu.be/MorvT9I0M9I) for Gemini API ^_^

ibigio commented on examples/Whisper_transcribe_device_microphone.ipynb:
```diff
@@ -0,0 +1,500 @@
+  {
+   "attachments": {
+    "Carl-AudioAPI-MicDemo.gif": {
+     "image/gif": "R0lGODlhIwPgAfdxACczRQA1igA5lwA8nAA/pgBAoQBDrAEyfQE5jgFNvQFRxwI1ggJWzQMxdgM8kgNZ1QRd2gZh4gdDmwkwawkybwpi3Atr8A0wZg9u7xFt7BIwXxIyYhI5dBRt5hUuWBVfyxcxWRdr4BguVRg+exkvUxorShoxVhsvUBszXBtEgBtm0xwvSxwxUhw3ZR0sRh0vSR0wSh0wTR02Xx4sQx4wRh4zVh8sQR8tQB8vRh8vRx8vSB84Yh87aR9s1yAsPiAsQCAvSSAwRyAyTiEvQiEvRSE1VyE6ZCE8ZSIsPCIwQyIySSI4XCM1UiQvQCQ4WSQ6XiQ8YCUvPCU3VSVs0CYzSCZBbiZEdig7WChKgihbqShk ...
```
you have two tables of contents, please remove one

ibigio replied:
tempfile is standard with python, don't need to install

ibigio replied:
```
error: You tried to access openai.Audio, but this is no longer supported in openai>=1.0.0 - see the README at https://github.com/openai/openai-python for the API.
```

gotta do:

```python
client.audio.transcriptions.create(...
```

ibigio replied:
this way of using the client is deprecated. do this instead:

```python
from openai import OpenAI

client = OpenAI(api_key="...")
```

ibigio replied:
ditto

ibigio commented on examples/Whisper_transcribe_device_microphone.ipynb:
```diff
@@ -0,0 +1,500 @@
+  {
+   "attachments": {
+    "Carl-AudioAPI-MicDemo.gif": {
+     "image/gif": "R0lGODlhIwPgAfdxACczRQA1igA5lwA8nAA/pgBAoQBDrAEyfQE5jgFNvQFRxwI1ggJWzQMxdgM8kgNZ1QRd2gZh4gdDmwkwawkybwpi3Atr8A0wZg9u7xFt7BIwXxIyYhI5dBRt5hUuWBVfyxcxWRdr4BguVRg+exkvUxorShoxVhsvUBszXBtEgBtm0xwvSxwxUhw3ZR0sRh0vSR0wSh0wTR02Xx4sQx4wRh4zVh8sQR8tQB8vRh8vRx8vSB84Yh87aR9s1yAsPiAsQCAvSSAwRyAyTiEvQiEvRSE1VyE6ZCE8ZSIsPCIwQyIySSI4XCM1UiQvQCQ4WSQ6XiQ8YCUvPCU3VSVs0CYzSCZBbiZEdig7WChKgihbqShk ...
```
In some notebooks (like vscode's) there's no way to get user/terminal input, so this gets stuck here. Potentially just record for a pre-determined amount of time.

CarlKho-Minerva replied:
Hi ibigio. It works on my vs code, but to accommodate notebooks that don't, I have made a parameter called `timed_recording (bool)` on the main/helper function. Appreciate you catching this, though. 
<img width="984" alt="image" src="https://github.com/user-attachments/assets/f6676a16-c04f-4310-92c0-3b39138b379e">

ibigio commented on examples/Whisper_transcribe_device_microphone.ipynb:
```diff
@@ -0,0 +1,500 @@
+  {
+   "attachments": {
+    "Carl-AudioAPI-MicDemo.gif": {
+     "image/gif": "R0lGODlhIwPgAfdxACczRQA1igA5lwA8nAA/pgBAoQBDrAEyfQE5jgFNvQFRxwI1ggJWzQMxdgM8kgNZ1QRd2gZh4gdDmwkwawkybwpi3Atr8A0wZg9u7xFt7BIwXxIyYhI5dBRt5hUuWBVfyxcxWRdr4BguVRg+exkvUxorShoxVhsvUBszXBtEgBtm0xwvSxwxUhw3ZR0sRh0vSR0wSh0wTR02Xx4sQx4wRh4zVh8sQR8tQB8vRh8vRx8vSB84Yh87aR9s1yAsPiAsQCAvSSAwRyAyTiEvQiEvRSE1VyE6ZCE8ZSIsPCIwQyIySSI4XCM1UiQvQCQ4WSQ6XiQ8YCUvPCU3VSVs0CYzSCZBbiZEdig7WChKgihbqShk ...
```
ditto

CarlKho-Minerva replied:
Hi @ibigio! Thanks for the response and review. 

Unfortunately, the ipynb files render as json and displays only line 3 - 6. 

Do you have a fix for this? It seems like a lot of people experience this too? https://github.com/orgs/community/discussions/12959

ibigio reviewed (changes requested):
| Criteria     | Description                                                                                         | Score |
| ------------ | --------------------------------------------------------------------------------------------------- | ----- |
| Relevance    | Is the content related to building with OpenAI technologies? Is it useful to others?                |     4  |
| Uniqueness   | Does the content offer new insights or unique information compared to existing documentation?       |    4   |
| Clarity      | Is the language easy to understand? Are things well-explained? Is the title clear?                  |    4   |
| Correctness  | Are the facts, code snippets, and examples correct and reliable? Does everything execute correctly? |     2  |
| Conciseness  | Is the content concise? Are all details necessary? Can it be made shorter?                          |     4  |
| Completeness | Is the content thorough and detailed? Are there things that weren’t explained fully?                |     4  |
| Grammar      | Are there grammatical or spelling errors present?                                                   |     4  |

Really solid contribution, thank you! Motivation is clear, steps are broken down well, and the sections make sense. Caught  a few mistakes here and there (mostly to do with using the SDK the old way), but once you correct them you're all set to merge!

CarlKho-Minerva committed:
feat: Heavily revise article for device microphone transcription in Whisper API

CarlKho-Minerva committed:
chore: update authors and registry yaml files

CarlKho-Minerva commented:
### Changelog

Hi @ibigio. Heavily revised my article now that I'm a month wiser. :)

**Updated Image:**
- Added `whisper_onChatGPTApp_cvk.gif` to the `images/` directory.

**Content Structure:**
- Removed duplicate table of contents.
- Replaced bolded numbered lists with H3 headers for cleaner formatting.

**Code Improvements:**
- Added translation functionality alongside transcription.
- Wrapped helper functions into separate functions for better modularity.
- Implemented `.env` for secure API key management.
- Specified data types for function parameters.
- Updated main function to include `is_english` parameter for language selection.
- Added timed recording option with `timed_recording` and `record_seconds` parameters.

**OpenAI API Updates:**
- Updated OpenAI library usage.
- Implemented `client.audio.translations.create` for non-English audio.

**Documentation:**
- Updated docstrings to reflect new functionality.
- Added additional demos for transcription and translation.
- Updated troubleshooting section and FAQ to cover new features.

**Terminology:**
- Updated text to reflect both "transcribe" and "translate" where appropriate.

**Aesthetic Improvements:**
- Enhanced overall formatting for better readability in VS Code.

CarlKho-Minerva committed:
chore: *correctly* update yaml files

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva commented:
@ibigio just a quick question—since it's my first time contributing: 

When you mentioned I could merge, does that mean there's an automatic process, or do I still need your review? I resolved the issues (including the dittos), so I thought it might enable merging. 

If not, could you review it when you have a moment? I know you're busy, so thanks a lot for your time!

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva commented:
> @ibigio just a quick question—since it's my first time contributing:
> 
> When you mentioned I could merge, does that mean there's an automatic process, or do I still need your review? I resolved the issues (including the dittos), so I thought it might enable merging.
> 
> If not, could you review it when you have a moment? I know you're busy, so thanks a lot for your time!

Hello again @ibigio! Gently bumping this up. ik ure busy so I understand if it takes a while. :)

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva committed:
indent fix

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva commented:
@gabor-openai @ericning-o @danielin-openai @ray-openai  hello folks! 

Perhaps @ibigio is caught up in the business amidst the good work he's doing for OAI. 

Gently tagging you guys so you could give the green signal in publishing this should it be satisfactory. 

Thank you very much for your hardwork!

ibigio commented on examples/Whisper_transcribe_device_microphone.ipynb:
Combining transcribing and translating here is a bit weird in this function, and also drops the `prompt` param for translations. (The prompt should be in english for translation and language of choice in a transcription). I'd split this out into two clear helper functions for translate and transcribe.

```python
def process_audio(file_name, is_english=True, prompt=""):
    with open(file_name, "rb") as audio_file:
        if is_english:
            response = client.audio.transcriptions.create(
                model="whisper-1", file=audio_file, prompt=prompt
            )
        else:
            response = client.audio.translations.create(
                model="whisper-1", file=audio_file
            )

        return response.text.strip()
```

ibigio commented on examples/Whisper_transcribe_device_microphone.ipynb:
I don't this this is how we intend for the prompt parameter to be used – looking at our [docs](https://platform.openai.com/docs/guides/speech-to-text#prompting), it is more of an example(s) than an instruction.
<img width="796" alt="Screenshot 2024-11-26 at 4 03 42 PM" src="https://github.com/user-attachments/assets/8bafbe38-203c-4d53-861b-b916f049cae7">

ibigio reviewed (changes requested):
Code seems to run now mostly free of errors, left a couple comments around code clarity and correctness.

CarlKho-Minerva commented:
Thanks @ibigio! Will get these fixed within an hour.

CarlKho-Minerva committed:
fix: clean up merged sections and remove conflict markers in registry.yaml

CarlKho-Minerva committed:
Merge branch 'carl-kho/Whisper_API-device_mic_transcription' of https://github.com/CarlKho-Minerva/openai-cookbook into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva committed:
feat: improve Whisper API prompts and refactor audio processing

- Refactor: Separate transcribe and translate functions
- Refactor: Clarify prompt usage in demos (example-based)
- Refactor: Add 5-second limit to Spanish translation demo
- Docs: Improve formatting and clarity of audio recording details
- Docs: Add note about prompt usage with links to API docs

CarlKho-Minerva commented:
Hi @ibigio,

Thanks again for your insightful feedback! I've addressed your comments in the latest commit:

-   **Prompts:** I've improved the prompts in both demos to be more example-based, providing context about the audio content, as suggested by the Whisper API documentation.
-   **Function Refactoring:** I've separated the `process_audio` function into `transcribe_audio_file` and `translate_audio_file` for better clarity and to ensure the `prompt` is used correctly in both cases.
-   **Time Limit:** I've added a 5-second time limit to the Spanish translation demo to prevent potential issues with indefinite recording.
-   **Documentation:** I've improved the formatting of the "Recording Audio Details" section, added a note about prompt usage with links to relevant documentation, and also restructured some of the installation steps to make them easier to follow.

Please take another look and let me know if you have any further suggestions!

P.S.: Sorry it took a while; there was a whirlwind of challenges following the [USCIS's policy change](https://washdiplomat.com/a-little-noticed-change-in-us-visa-policy-threatens-global-education/) that threatened my ability to study in the US. It is resolved now,t though.

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva committed:
Merge branch 'main' into carl-kho/Whisper_API-device_mic_transcription

CarlKho-Minerva commented:
My academic focus has shifted to general computer (Omniparser v2) use with [Inner Speech](https://www.nature.com/articles/s41597-022-01147-2). 

Is there anyone on the team I can get on a 15-min call for starting-out resources? Would appreciate links as well.

Also, gently bumping up @ibigio. Hope everything's well over there.

CarlKho-Minerva commented:
Bumping to remove stale. :)

Djubz reviewed (commented):
Updated registry.yaml ;)

CarlKho-Minerva commented:
Bumping to remove stale. I finished my internship at Dell Tech now. Distributed systems is fun.

CarlKho-Minerva commented:
Focusing on [BCI](https://somach.vercel.app/) for gaming now!

LeoKas85 commented:
It’s going down down down let’s get this over with!

CarlKho-Minerva commented:
replicating MIT's $1250 (2018) Subvocalization tech [for $30](https://somach.vercel.app/editorial/jump.html) now

CarlKho-Minerva commented:
I just finished my capstone final thesis and passed! https://somach.vercel.app/somach/final-capstone-slides.html

CarlKho-Minerva commented:
so i have 2 c-corp startups now: https://somach.life/ & https://themildlyusefulcompany.com/
